pkg/controllers/managedzone/managedzone_controller.go

```diff
@@ -0,0 +1,371 @@
+		} else {
+			return ctrl.Result{}, err
+		}
+	}
+	managedZone := previous.DeepCopy()
+
+	log.Log.V(3).Info("ManagedZoneReconciler Reconcile", "managedZone", managedZone)
+
+	if managedZone.DeletionTimestamp != nil && !managedZone.DeletionTimestamp.IsZero() {
+		if err := r.deleteParentZoneNSRecord(ctx, managedZone); err != nil {
+			log.Log.Error(err, "Failed to delete parent Zone NS Record", "managedZone", managedZone)
+			return ctrl.Result{}, err
+		}
+		if err := r.deleteManagedZone(ctx, managedZone); err != nil {
+			log.Log.Error(err, "Failed to delete ManagedZone", "managedZone", managedZone)
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(managedZone, ManagedZoneFinalizer)
+
+		err = r.Update(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(managedZone, ManagedZoneFinalizer) {
+		log.Log.Info("ManagedZoneReconciler Reconcile: Added finalizer and set parent zone owner", "managedZone", managedZone.Name)
+
+		controllerutil.AddFinalizer(managedZone, ManagedZoneFinalizer)
+
+		err = r.setParentZoneOwner(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = r.Update(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	var reason, message string
+	status := v1.ConditionTrue
+
+	// Publish the managed zone
+	err = r.publishManagedZone(ctx, managedZone)
+	if err != nil {
+		status = v1.ConditionFalse
+		reason = "ProviderError"
+		message = fmt.Sprintf("The DNS provider failed to ensure the managed zone: %v", err)
+
+		err = r.Status().Update(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Create the parent zone NS record
+	err = r.createParentZoneNSRecord(ctx, managedZone)
+	if err != nil {
+		status = v1.ConditionFalse
+		reason = "ParentZoneNSRecordError"
+		message = fmt.Sprintf("Failed to create the NS record in the parent managed zone: %v", err)
+
+		err = r.Status().Update(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Check the parent zone NS record status
+	err = r.parentZoneNSRecordReady(ctx, managedZone)
+	if err != nil {
+		status = v1.ConditionFalse
+		reason = "ParentZoneNSRecordNotReady"
+		message = fmt.Sprintf("NS Record ready status check failed: %v", err)
+
+		err = r.Status().Update(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	managedZone.Status.ObservedGeneration = managedZone.Generation
+	setManagedZoneCondition(managedZone, v1.ManagedZoneConditionReady, status, reason, message)
+	err = r.Status().Update(ctx, managedZone)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ManagedZoneReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.ManagedZone{}).
+		Owns(&v1.DNSRecord{}).
```

Comment:
Interesting does this trigger a reconcile if something it owns changes? Do we need to own and for ManagedZone?